aten/src/ATen/native/cuda/SortingRadixSelect.cuh

```diff
@@ -213,9 +216,13 @@ __device__ void countRadixUsingMask(
 #if defined(USE_ROCM)
       counts[j] += __popcll(WARP_BALLOT(vote));
 #else
-      counts[j] += __popc(WARP_BALLOT(vote, ACTIVE_MASK()));
+      counts[j] += __popc(WARP_BALLOT(vote, mask));
 #endif
     }
+    i += blockDim.x;
+#if !defined(USE_ROCM)
+    mask = WARP_BALLOT(i < sliceSize, mask);
```

Comment:
is this needlessly complicated? `mask` can be not `0xffffffff` only on the last iteration, where updating it for the next iteration is pointless?

Reply:
Assuming only the last iteration will result in a subset of threads exiting the loop early, then `mask` will be used by the subset still in the loop to conduct the ballot. Updating the `mask` on the last iteration is unnecessary, but how is one thread supposed to figure out that it's the last iteration? That's the question I had since I wasn't sure you could rely on this assumption. Is there a scenario where a subset of threads do multiple iterations?

If the assumption holds, then a more efficient solution would be to precompute the number of iterations and for iterations - 1 use a ballot for all threads, while on the last iteration, there would be two ballots. One to determine what threads are still in the iteration, and another to do the normal ballot. The overhead for the extra ballots would be drastically reduced to just one extra ballot instead of one on every iteration.

Reply:
That's right, you could avoid most warp ballots this way but that requires bigger code changes. Without that, could you just leave the code as it was and use `WARP_BALLOT(i < sliceSize)` instead of `ACTIVE_MASK()`? You never need to and predicate with mask I think.

Reply:
> [C]ould you just leave the code as it was and use `WARP_BALLOT(i < sliceSize)` instead of `ACTIVE_MASK()`?

No, you can't drop the `mask` since implicitly a full mask will be used. In the case where not all threads can reach a primitive with a full mask, then the program could hang. I got this from [this NVIDIA blogpost](https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/) under Update Legacy Warp-Level Programming point 1.

Reply:
cool, can you leave a comment for the first WARP_BALLOT call then, that it needs to be called outside the loop to make sure all threads participate?

Reply:
I have added the comment. If you want, I can draft a proposal which just does an extra warp ballot for the last iteration. 

I was thinking it'd look like this somehow

```cuda
index_t iters = ceil_div<index_t>(sliceSize, (index_t) blockDim.x);
index_t i = threadIdx.x,;
for (index_t k = 0; k < iters - 1; i += blockDim.x, k++) {
   ...
   counts[j] += __popc(WARP_BALLOT(vote));
   ...
}
i += blockDim.x;
unsigned mask = WARP_BALLOT(i < sliceSize);
if (i  < sliceSize) {
  ...
  counts[j] += __popc(WARP_BALLOT(vote, mask));
  ...
}
```

Reply:
It's fine, I don't think this kernel is perf-critical